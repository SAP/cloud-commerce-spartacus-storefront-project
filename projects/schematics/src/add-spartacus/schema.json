{
  "$schema": "http://json-schema.org/schema",
  "id": "SpartacusSchematics",
  "title": "Spartacus Schematics",
  "type": "object",
  "properties": {
    "project": {
      "type": "string",
      "description": "The name of the project.",
      "$default": {
        "$source": "projectName"
      }
    },
    "baseUrl": {
      "type": "string",
      "description": "The url of the OCC backend",
      "default": "https://localhost:9002"
    },
    "baseSite": {
      "type": "string",
      "description": "The base site to use with spartacus",
      "default": "electronics"
    },
    "useMetaTags": {
      "type": "boolean",
      "description": "Configure baseUrl via meta tag in index.html"
    },
    "featureLevel": {
      "type": "string",
      "description": "Configure feature level for application.",
      "default": "1.1"
    },
    "overwriteAppComponent": {
      "type": "boolean",
      "description": "Overwrite whole app.component.html file.",
      "default": false
    }
  },
  "required": []
}
