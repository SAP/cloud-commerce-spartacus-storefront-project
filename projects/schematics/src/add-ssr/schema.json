{
  "$schema": "http://json-schema.org/schema",
  "id": "SpartacusSchematics",
  "title": "Spartacus Schematics",
  "type": "object",
  "properties": {
    "project": {
      "type": "string",
      "description": "The name of the project.",
      "$default": {
        "$source": "projectName"
      }
    }
  },
  "required": []
}
