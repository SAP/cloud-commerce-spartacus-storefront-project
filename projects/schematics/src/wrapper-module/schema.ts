import { ExecutionOptions } from '@angular-devkit/schematics';

export interface Schema extends Partial<ExecutionOptions> {
  /**
   * The name of the project in which to execute schematics.
   */
  project: string;
  /**
   * When enabled, prints the additional logs.
   */
  debug?: boolean;
  /**
   * The marker module for which
   * to find or generate a new wrapper module.
   */
  markerModuleName: string;
  /**
   * Name of the feature module
   * to add to the wrapper module.
   */
  featureModuleName: string;
  /**
   * Internal options.
   * Should not be set by the user.
   */
  internal?: {
    /**
     * The execution sequence.
     * Used for ensuring one log message is printed.
     */
    sequence?: number;
  };
}
