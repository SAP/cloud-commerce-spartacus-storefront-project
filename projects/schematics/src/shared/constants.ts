export const UTF_8 = 'utf-8';

/***** Imports start *****/
export const ANGULAR_SCHEMATICS = '@schematics/angular';
export const ANGULAR_CORE = '@angular/core';
export const ANGULAR_FORMS = '@angular/forms';
export const ANGULAR_ROUTER = '@angular/router';
export const RXJS = 'rxjs';

export const SPARTACUS_CORE = '@spartacus/core';
export const SPARTACUS_STOREFRONTLIB = '@spartacus/storefront';
export const SPARTACUS_CDS = '@spartacus/cds';

export const NGRX_STORE = '@ngrx/store';
/***** Imports end *****/

/***** Classes start *****/
export const OBSERVABLE_CLASS = 'Observable';

export const STORE = 'Store';

export const FORM_BUILDER = 'FormBuilder';

export const CHANGE_DETECTOR_REF = 'ChangeDetectorRef';

export const ACTIVATED_ROUTE = 'ActivatedRoute';
export const ROUTER = 'Router';

export const CMS_COMPONENT_DATA_CLASS = 'CmsComponentData';
export const CONFIG_MODULE_CLASS = 'ConfigModule';
export const CMS_CONFIG = 'CmsConfig';
export const CMS_SELECTORS = 'CmsSelectors';
export const CMS_ACTIONS = 'CmsActions';
export const LOAD_CMS_COMPONENT_CLASS = 'LoadCmsComponent';
export const LOAD_CMS_COMPONENT_FAIL_CLASS = 'LoadCmsComponentFail';
export const LOAD_CMS_COMPONENT_SUCCESS_CLASS = 'LoadCmsComponentSuccess';
export const CMS_GET_COMPONENT_FROM_PAGE = 'CmsGetComponentFromPage';
export const USER_ADDRESS_SERVICE = 'UserAddressService';
export const AUTH_SERVICE = 'AuthService';
export const FEATURE_CONFIG_SERVICE = 'FeatureConfigService';
export const PAGE_META_RESOLVER = 'PageMetaResolver';
export const CMS_SERVICE = 'CmsService';
export const PAGE_META_SERVICE = 'PageMetaService';
export const CHECKOUT_SERVICE = 'CheckoutService';
export const CHECKOUT_PAYMENT_SERVICE = 'CheckoutPaymentService';
export const CHECKOUT_DELIVERY_SERVICE = 'CheckoutDeliveryService';
export const CART_DATA_SERVICE = 'CartDataService';
export const CART_SERVICE = 'CartService';
export const ACTIVE_CART_SERVICE = 'ActiveCartService';
export const PROMOTION_SERVICE = 'PromotionService';
export const ORDER_DETAILS_SERVICE = 'OrderDetailsService';
export const CHECKOUT_LOGIN_COMPONENT = 'CheckoutLoginComponent';
export const AUTH_REDIRECT_SERVICE = 'AuthRedirectService';
export const CHECKOUT_DETAILS_SERVICE = 'CheckoutDetailsService';
export const NOT_CHECKOUT_AUTH_GUARD = 'NotCheckoutAuthGuard';
export const ROUTING_SERVICE = 'RoutingService';
export const SHIPPING_ADDRESS_COMPONENT = 'ShippingAddressComponent';
export const CHECKOUT_CONFIG_SERVICE = 'CheckoutConfigService';
export const TRANSLATION_SERVICE = 'TranslationService';
export const CHECKOUT_PAGE_META_RESOLVER = 'CheckoutPageMetaResolver';
export const ADD_TO_CART_COMPONENT = 'AddToCartComponent';
export const MODAL_SERVICE = 'ModalService';
export const CURRENT_PRODUCT_SERVICE = 'CurrentProductService';
export const CART_NOT_EMPTY_GUARD = 'CartNotEmptyGuard';
export const CART_TOTALS_COMPONENT = 'CartTotalsComponent';
export const MINI_CART_COMPONENT = 'MiniCartComponent';
export const CHECKOUT_ORDER_SUMMARY_COMPONENT = 'CheckoutOrderSummaryComponent';
export const CHECKOUT_PROGRESS_MOBILE_TOP_COMPONENT =
  'CheckoutProgressMobileTopComponent';
export const CHECKOUT_CONFIG = 'CheckoutConfig';
export const ROUTING_CONFIG_SERVICE = 'RoutingConfigService';
export const PAYMENT_METHOD_COMPONENT = 'PaymentMethodComponent';
export const USER_PAYMENT_SERVICE = 'UserPaymentService';
export const GLOBAL_MESSAGE_SERVICE = 'GlobalMessageService';
export const CHECKOUT_AUTH_GUARD = 'CheckoutAuthGuard';
export const CART_PAGE_LAYOUT_HANDLER = 'CartPageLayoutHandler';
export const SELECTIVE_CART_SERVICE = 'SelectiveCartService';
export const CDS_SPARTACUS_EVENT_SERVICE = 'SpartacusEventService';
export const CONSENT_SERVICE = 'ConsentService';
export const CDS_CONFIG = 'CdsConfig';
export const ADDRESS_BOOK_COMPONENT_SERVICE = 'AddressBookComponentService';
export const CHECKOUT_GUARD = 'CheckoutGuard';
export const EXPRESS_CHECKOUT_SERVICE = 'ExpressCheckoutService';
export const CONSENT_MANAGEMENT_FORM_COMPONENT =
  'ConsentManagementFormComponent';
/***** Classes end *****/

/***** Removed public api start *****/
export const CART_EFFECTS = 'CartEffects';
export const WISHLIST_EFFECTS = 'WishlistEffects';
export const CART_VOUCHER_EFFECTS = 'CartVoucherEffects';
export const CART_ENTRY_EFFECTS = 'CartEntryEffects';
export const CART_COMBINED_EFFECTS = 'effects';

export const GET_REDUCERS = 'getReducers';
export const REDUCER_TOKEN = 'reducerToken';
export const REDUCER_PROVIDER = 'reducerProvider';
export const CLEAR_CART_STATE = 'clearCartState';
export const META_REDUCERS = 'metaReducers';
export const CLEAR_MULTI_CART_STATE = 'clearMultiCartState';
export const MULTI_CART_META_REDUCERS = 'multiCartMetaReducers';
export const MULTI_CART_REDUCER_TOKEN = 'multiCartReducerToken';
export const GET_MULTI_CART_REDUCERS = 'getMultiCartReducers';
export const MULTI_CART_REDUCER_PROVIDER = 'multiCartReducerProvider';

export const SITE_CONTEXT_PARAMS_PROVIDERS = 'siteContextParamsProviders';
export const INIT_SITE_CONTEXT_ROUTES_HANDLER = 'initSiteContextRoutesHandler';
export const INITITIALIZE_CONTEXT = 'inititializeContext';
export const CONTEXT_SERVICE_PROVIDERS = 'contextServiceProviders';
/***** Removed public api end *****/

/***** Properties start *****/
export const CMS_COMPONENT_DATA_PROPERTY_NAME = 'componentData';
/***** Properties end *****/

/***** APIs start *****/
export const TODO_SPARTACUS = 'TODO:Spartacus -';

export const GET_COMPONENT_STATE_OLD_API = 'getComponentState';
export const GET_COMPONENTS_STATE_NEW_API = 'getComponentsState';

export const GET_COMPONENT_ENTITIES_OLD_API = 'getComponentEntities';
export const GET_COMPONENT_ENTITIES_COMMENT = `// ${TODO_SPARTACUS} '${GET_COMPONENT_ENTITIES_OLD_API}' has been removed, please use some of the newer API methods.`;

export const COMPONENT_STATE_SELECTOR_FACTORY_OLD_API =
  'componentStateSelectorFactory';
export const COMPONENTS_STATE_SELECTOR_FACTORY_NEW_API =
  'componentsLoaderStateSelectorFactory';

export const COMPONENT_SELECTOR_FACTORY_OLD_API = 'componentSelectorFactory';
export const COMPONENTS_SELECTOR_FACTORY_NEW_API = 'componentsSelectorFactory';
/***** APIs end *****/

/***** Misc start *****/
export const CSS_DOCS_URL =
  'https://sap.github.io/cloud-commerce-spartacus-storefront-docs/deprecation-guide/css';
/***** Misc end *****/
