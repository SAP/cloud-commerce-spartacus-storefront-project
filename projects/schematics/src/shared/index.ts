export * from './constants';
export * from './feature-libs-constants';
export * from './utils/index';
