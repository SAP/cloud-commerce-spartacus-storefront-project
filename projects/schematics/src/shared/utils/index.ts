export {
  createNewConfig,
  getConfig,
  getExistingStorefrontConfigNode,
  mergeConfig,
} from './config-utils';
export {
  commitChanges,
  findMultiLevelNodesByTextAndKind,
  getTsSourceFile,
  InsertDirection,
  removeImport,
} from './file-utils';
export {
  addImport,
  addToMetadata,
  addToModuleImports,
  addToModuleImportsAndCommitChanges,
  addToModuleProviders,
  createImportChange,
} from './module-file-utils';
export {
  getSpartacusSchematicsVersion,
  readPackageJson,
} from './package-utils';
export {
  getDefaultProjectNameFromWorkspace,
  getProjectTargets,
  getSourceRoot,
  getWorkspace,
} from './workspace-utils';
