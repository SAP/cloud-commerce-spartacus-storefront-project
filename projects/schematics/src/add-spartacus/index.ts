import { experimental, JsonParseMode, parseJson } from '@angular-devkit/core';
import { italic, red } from '@angular-devkit/core/src/terminal';
import {
  chain,
  noop,
  Rule,
  SchematicContext,
  SchematicsException,
  Tree,
} from '@angular-devkit/schematics';
import { NodePackageInstallTask } from '@angular-devkit/schematics/tasks';
import {
  appendHtmlElementToHead,
  getProjectStyleFile,
  getProjectTargetOptions,
} from '@angular/cdk/schematics';
import {
  addSymbolToNgModuleMetadata,
  insertImport,
  isImported,
} from '@schematics/angular/utility/ast-utils';
import { InsertChange } from '@schematics/angular/utility/change';
import {
  addPackageJsonDependency,
  NodeDependency,
  NodeDependencyType,
} from '@schematics/angular/utility/dependencies';
import { getAppModulePath } from '@schematics/angular/utility/ng-ast-utils';
import { getProjectTargets } from '@schematics/angular/utility/project-targets';
import { Schema as SpartacusOptions } from './schema';
import { getTsSourceFile } from '../shared/utils/file-utils';

function getWorkspace(
  host: Tree
): { path: string; workspace: experimental.workspace.WorkspaceSchema } {
  const possibleFiles = ['/angular.json', '/.angular.json'];
  const path = possibleFiles.filter(filePath => host.exists(filePath))[0];

  if (!path) {
    throw new SchematicsException(`Could not find Angular`);
  }

  const configBuffer = host.read(path);
  if (configBuffer === null) {
    throw new SchematicsException(`Could not find (${path})`);
  }
  const content = configBuffer.toString();

  return {
    path,
    workspace: (parseJson(
      content,
      JsonParseMode.Loose
    ) as {}) as experimental.workspace.WorkspaceSchema,
  };
}

function addPackageJsonDependencies(): Rule {
  const spartacusVersion = '^1.0.0';
  const ngrxVersion = '^8.0.0';

  return (tree: Tree, context: SchematicContext) => {
    const dependencies: NodeDependency[] = [
      {
        type: NodeDependencyType.Default,
        version: spartacusVersion,
        name: '@spartacus/core',
      },
      {
        type: NodeDependencyType.Default,
        version: spartacusVersion,
        name: '@spartacus/storefront',
      },
      {
        type: NodeDependencyType.Default,
        version: spartacusVersion,
        name: '@spartacus/assets',
      },
      {
        type: NodeDependencyType.Default,
        version: spartacusVersion,
        name: '@spartacus/styles',
      },

      {
        type: NodeDependencyType.Default,
        version: '^4.1.0',
        name: '@ng-bootstrap/ng-bootstrap',
      },
      {
        type: NodeDependencyType.Default,
        version: '^2.13.2',
        name: '@ng-select/ng-select',
      },

      {
        type: NodeDependencyType.Default,
        version: ngrxVersion,
        name: '@ngrx/store',
      },
      {
        type: NodeDependencyType.Default,
        version: ngrxVersion,
        name: '@ngrx/effects',
      },
      {
        type: NodeDependencyType.Default,
        version: ngrxVersion,
        name: '@ngrx/router-store',
      },

      {
        type: NodeDependencyType.Default,
        version: '^4.2.1',
        name: 'bootstrap',
      },
      { type: NodeDependencyType.Default, version: '^15.0.6', name: 'i18next' },
      {
        type: NodeDependencyType.Default,
        version: '^2.0.1',
        name: 'i18next-xhr-backend',
      },
    ];

    dependencies.forEach(dependency => {
      addPackageJsonDependency(tree, dependency);
      context.logger.log(
        'info',
        `✅️ Added '${dependency.name}' into ${dependency.type}`
      );
    });

    return tree;
  };
}

function installPackageJsonDependencies(): Rule {
  return (tree: Tree, context: SchematicContext) => {
    context.addTask(new NodePackageInstallTask());
    context.logger.log('info', `🔍 Installing packages...`);
    return tree;
  };
}

function addImport(
  host: Tree,
  modulePath: string,
  importText: string,
  importPath: string
) {
  const moduleSource = getTsSourceFile(host, modulePath) as any;
  if (!isImported(moduleSource, importText, importPath)) {
    const change = insertImport(
      moduleSource,
      modulePath,
      importText,
      importPath
    );
    if (change) {
      const recorder = host.beginUpdate(modulePath);
      recorder.insertLeft(
        (change as InsertChange).pos,
        (change as InsertChange).toAdd
      );
      host.commitUpdate(recorder);
    }
  }
}

function importModule(host: Tree, modulePath: string, importText: string) {
  const moduleSource = getTsSourceFile(host, modulePath);
  const metadataChanges = addSymbolToNgModuleMetadata(
    moduleSource,
    modulePath,
    'imports',
    importText
  );
  if (metadataChanges) {
    const recorder = host.beginUpdate(modulePath);
    metadataChanges.forEach((change: InsertChange) => {
      recorder.insertRight(change.pos, change.toAdd);
    });
    host.commitUpdate(recorder);
  }
}

function getStorefrontConfig(options: SpartacusOptions): string {
  const baseUrlPart = `baseUrl: '${options.baseUrl}',`;
  return `{
      backend: {
        occ: {${options.useMetaTags ? '' : baseUrlPart}
          prefix: '/rest/v2/'
        }
      },
      authentication: {
        client_id: 'mobile_android',
        client_secret: 'secret'
      },
      context: {
        baseSite: ['${options.baseSite}']
      },
      i18n: {
        resources: translations,
        chunks: translationChunksConfig,
        fallbackLang: 'en'
      },
      features: {
        level: '${options.featureLevel}'
      }
    }`;
}

function updateAppModule(options: SpartacusOptions): Rule {
  return (host: Tree, context: SchematicContext) => {
    context.logger.debug('Updating main module');

    // find app module
    const projectTargets = getProjectTargets(host, options.project);

    if (!projectTargets.build) {
      throw new SchematicsException(`Project target "build" not found.`);
    }

    const mainPath = projectTargets.build.options.main;
    const modulePath = getAppModulePath(host, mainPath);
    context.logger.debug(`main module path: ${modulePath}`);

    // add imports
    addImport(host, modulePath, 'translations', '@spartacus/assets');
    addImport(host, modulePath, 'translationChunksConfig', '@spartacus/assets');
    addImport(host, modulePath, 'B2cStorefrontModule', '@spartacus/storefront');
    addImport(
      host,
      modulePath,
      'defaultCmsContentConfig',
      '@spartacus/storefront'
    );
    addImport(host, modulePath, 'ConfigModule', '@spartacus/core');

    importModule(
      host,
      modulePath,
      `B2cStorefrontModule.withConfig(${getStorefrontConfig(options)})`
    );
    importModule(
      host,
      modulePath,
      `ConfigModule.withConfigFactory(defaultCmsContentConfig)`
    );

    return host;
  };
}

function installStyles(project: experimental.workspace.WorkspaceProject): Rule {
  return (host: Tree) => {
    const styleFilePath = getProjectStyleFile(project);

    if (!styleFilePath) {
      console.warn(
        red(`Could not find the default style file for this project.`)
      );
      console.warn(red(`Please consider manually setting up spartacus styles`));
      return;
    }

    if (styleFilePath.split('.').pop() !== 'scss') {
      console.warn(
        red(`Could not find the default SCSS style file for this project. `)
      );
      console.warn(
        red(
          `Please make sure your project is configured with SCSS and consider manually setting up spartacus styles.`
        )
      );
      return;
    }

    const buffer = host.read(styleFilePath);

    if (!buffer) {
      console.warn(
        red(
          `Could not read the default style file within the project ` +
            `(${italic(styleFilePath)})`
        )
      );
      console.warn(red(`Please consider manually importing spartacus styles.`));
      return;
    }

    const htmlContent = buffer.toString();
    const insertion = '\n' + `@import '~@spartacus/styles/index';\n`;

    if (htmlContent.includes(insertion)) {
      return;
    }

    const recorder = host.beginUpdate(styleFilePath);

    recorder.insertLeft(htmlContent.length, insertion);
    host.commitUpdate(recorder);
  };
}

function updateMainComponent(
  project: experimental.workspace.WorkspaceProject,
  options: SpartacusOptions
): Rule {
  return (host: Tree, _context: SchematicContext) => {
    const filePath = project.sourceRoot + '/app/app.component.html';
    const buffer = host.read(filePath);

    if (!buffer) {
      console.warn(red(`Could not read app.component.html file.`));
      return;
    }

    const htmlContent = buffer.toString();
    const insertion = '\n' + `<cx-storefront></cx-storefront>\n`;

    if (htmlContent.includes(insertion)) {
      return;
    }

    const recorder = host.beginUpdate(filePath);

    if (options && options.overwriteAppComponent) {
      recorder.remove(0, htmlContent.length);
      recorder.insertLeft(0, insertion);
    } else {
      recorder.insertLeft(htmlContent.length, insertion);
    }

    host.commitUpdate(recorder);

    return host;
  };
}

export function getIndexHtmlPath(
  project: experimental.workspace.WorkspaceProject
): string {
  const buildOptions = getProjectTargetOptions(project, 'build');

  if (!buildOptions.index) {
    throw new SchematicsException('"index.html" file not found.');
  }

  return buildOptions.index;
}

function updateIndexFile(
  project: experimental.workspace.WorkspaceProject,
  options: SpartacusOptions
): Rule {
  return (host: Tree) => {
    const projectIndexHtmlPath = getIndexHtmlPath(project);
    const baseUrl = options.baseUrl || 'OCC_BACKEND_BASE_URL_VALUE';

    const metaTags = [
      `<meta name="occ-backend-base-url" content="${baseUrl}" />`,
      `<meta name="media-backend-base-url" content="MEDIA_BACKEND_BASE_URL_VALUE" />`,
    ];

    metaTags.forEach(metaTag => {
      appendHtmlElementToHead(host, projectIndexHtmlPath, metaTag);
    });

    return host;
  };
}

export function addSpartacus(options: SpartacusOptions): Rule {
  return (tree: Tree, context: SchematicContext) => {
    const { workspace } = getWorkspace(tree);

    if (!options.project) {
      throw new SchematicsException('Option "project" is required.');
    }

    const project = workspace.projects[options.project];
    if (!project) {
      throw new SchematicsException(
        `Project is not defined in this workspace.`
      );
    }

    if (project.projectType !== 'application') {
      throw new SchematicsException(
        `Spartacus requires a project type of "application".`
      );
    }

    return chain([
      addPackageJsonDependencies(),
      updateAppModule(options),
      installStyles(project),
      updateMainComponent(project, options),
      options.useMetaTags ? updateIndexFile(project, options) : noop(),
      installPackageJsonDependencies(),
    ])(tree, context);
  };
}
