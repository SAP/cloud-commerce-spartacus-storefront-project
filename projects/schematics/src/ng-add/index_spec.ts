import {
  SchematicTestRunner,
  UnitTestTree,
} from '@angular-devkit/schematics/testing';
import * as path from 'path';
import { Schema as SpartacusOptions } from '../add-spartacus/schema';
import { NGUNIVERSAL_EXPRESS_ENGINE, UTF_8 } from '../shared/constants';
import { getPathResultsForFile } from '../shared/utils/file-utils';

const collectionPath = path.join(__dirname, '../collection.json');

// tslint:disable:max-line-length
describe('Spartacus Schematics: ng-add', () => {
  const schematicRunner = new SchematicTestRunner('schematics', collectionPath);

  let appTree: UnitTestTree;

  const workspaceOptions: any = {
    name: 'workspace',
    version: '0.5.0',
  };

  const appOptions: any = {
    name: 'schematics-test',
    inlineStyle: false,
    inlineTemplate: false,
    routing: false,
    style: 'scss',
    skipTests: false,
    projectRoot: '',
  };

  const defaultOptions: SpartacusOptions = {
    project: 'schematics-test',
  };

  beforeEach(async () => {
    appTree = await schematicRunner
      .runExternalSchematicAsync(
        '@schematics/angular',
        'workspace',
        workspaceOptions
      )
      .toPromise();
    appTree = await schematicRunner
      .runExternalSchematicAsync(
        '@schematics/angular',
        'application',
        appOptions,
        appTree
      )
      .toPromise();
  });

  it('should add spartacus deps', async () => {
    const tree = await schematicRunner
      .runSchematicAsync(
        'ng-add',
        { ...defaultOptions, name: 'schematics-test' },
        appTree
      )
      .toPromise();

    // Verify that we registered task only for installing packages (no PWA schematics)
    const { tasks } = schematicRunner;
    expect(tasks[0].name).toEqual('node-package');
    expect((tasks[0].options as any).command).toEqual('install');
    expect(tasks.length).toBe(1);

    const packageJson = tree.readContent('/package.json');
    const packageObj = JSON.parse(packageJson);
    const depPackageList = Object.keys(packageObj.dependencies);
    expect(depPackageList.includes('@spartacus/core')).toBe(true);
    expect(depPackageList.includes('@spartacus/storefront')).toBe(true);
    expect(depPackageList.includes('@spartacus/styles')).toBe(true);
  });

  it('should add spartacus with PWA via passed parameter', async () => {
    await schematicRunner
      .runSchematicAsync(
        'ng-add',
        { ...defaultOptions, name: 'schematics-test', pwa: true },
        appTree
      )
      .toPromise();

    // Verify that we registered task for installing and running @angular/pwa schematics
    const { tasks } = schematicRunner;
    // We have 2 node-package install tasks registered
    // Order matters. First install needs to be executed and then angular-pwa schematics
    expect(tasks[1].name).toEqual('node-package');
    expect((tasks[1].options as any).command).toEqual('install');
    expect(tasks[2].name).toEqual('run-schematic');
    expect((tasks[2].options as any).name).toEqual(
      'run-angular-pwa-schematics'
    );

    // Run @angular/pwa schematics as it was registered as future task
    const tree = await schematicRunner
      .runSchematicAsync('run-angular-pwa-schematics', defaultOptions, appTree)
      .toPromise();

    const buffer = tree.read('src/manifest.webmanifest');
    expect(buffer).toBeTruthy();

    if (buffer) {
      const webmanifestJSON = JSON.parse(buffer.toString(UTF_8));
      expect(webmanifestJSON.name).toEqual(defaultOptions.project);
    }
  });

  it('should add spartacus with SSR via passed parameter', async () => {
    const tree = await schematicRunner
      .runSchematicAsync(
        'ng-add',
        { ...defaultOptions, name: 'schematics-test', ssr: true },
        appTree
      )
      .toPromise();
    const packageJsonBuffer = tree.read('/package.json');
    expect(packageJsonBuffer).toBeTruthy();
    const appServerModulePath = getPathResultsForFile(
      tree,
      'app.server.module.ts',
      '/src'
    )[0];
    const appServerModuleBuffer = tree.read(appServerModulePath);
    expect(appServerModuleBuffer).toBeTruthy();
    const serverBuffer = tree.read('server.ts');
    expect(serverBuffer).toBeTruthy();

    if (packageJsonBuffer) {
      const packageJSON = JSON.parse(packageJsonBuffer.toString(UTF_8));
      expect(packageJSON.dependencies[NGUNIVERSAL_EXPRESS_ENGINE]).toBeTruthy();
      expect(packageJSON.dependencies['@angular/platform-server']).toBeTruthy();
    }

    if (appServerModuleBuffer) {
      const appServerModuleContent = appServerModuleBuffer.toString(UTF_8);
      expect(
        appServerModuleContent.includes('ServerTransferStateModule')
      ).toBeTruthy();
    }
  });

  describe('@angular/localize', () => {
    it('should provide import in polyfills.ts and main.server.ts if SSR enabled', async () => {
      const tree = await schematicRunner
        .runSchematicAsync(
          'ng-add',
          { ...defaultOptions, name: 'schematics-test', ssr: true },
          appTree
        )
        .toPromise();

      const polyfillsPath = getPathResultsForFile(
        appTree,
        'polyfills.ts',
        '/src'
      )[0];

      const buffer = tree.read('./server.ts');
      const polyfillsBuffer = tree.read(polyfillsPath);
      expect(buffer).toBeTruthy();
      expect(polyfillsBuffer).toBeTruthy();
      if (buffer) {
        const appServerTsFileString = buffer.toString(UTF_8);
        expect(
          appServerTsFileString.includes("import '@angular/localize/init'")
        ).toBeTruthy();
      }
      if (polyfillsBuffer) {
        const polyfills = polyfillsBuffer.toString(UTF_8);
        expect(
          polyfills.includes("import '@angular/localize/init'")
        ).toBeTruthy();
      }
    });
  });
});
