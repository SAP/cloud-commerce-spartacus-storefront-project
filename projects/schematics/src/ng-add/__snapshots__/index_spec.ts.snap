// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Spartacus Schematics: ng-add should add spartacus properly with SSR 1`] = `
"import { HttpClientModule } from "@angular/common/http";
import { NgModule } from '@angular/core';
import { BrowserModule } from '@angular/platform-browser';
import { EffectsModule } from "@ngrx/effects";
import { StoreModule } from "@ngrx/store";
import { AppRoutingModule } from "@spartacus/storefront";
import { AppComponent } from './app.component';
import { SpartacusModule } from './spartacus/spartacus.module';

@NgModule({
  declarations: [
    AppComponent
  ],
  imports: [
    BrowserModule,
    HttpClientModule,
    AppRoutingModule,
    StoreModule.forRoot({}),
    EffectsModule.forRoot([]),
    SpartacusModule
  ],
  providers: [],
  bootstrap: [AppComponent]
})
export class AppModule { }
"
`;

exports[`Spartacus Schematics: ng-add should add spartacus properly with SSR 2`] = `
"{
  "name": "workspace",
  "version": "0.0.0",
  "scripts": {
    "ng": "ng",
    "start": "ng serve",
    "build": "ng build",
    "watch": "ng build --watch --configuration development",
    "test": "ng test",
    "dev:ssr": "ng run schematics-test:serve-ssr",
    "serve:ssr": "node dist/schematics-test/server/main.js",
    "build:ssr": "ng build && ng run schematics-test:server",
    "prerender": "ng run schematics-test:prerender"
  },
  "private": true,
  "dependencies": {
    "@angular/animations": "^16.2.0",
    "@angular/common": "^16.2.0",
    "@angular/compiler": "^16.2.0",
    "@angular/core": "^16.2.0",
    "@angular/forms": "^16.2.0",
    "@angular/platform-browser": "^16.2.0",
    "@angular/platform-browser-dynamic": "^16.2.0",
    "@angular/platform-server": "^16.2.10",
    "@angular/router": "^16.2.0",
    "@angular/service-worker": "^16.2.10",
    "@fontsource/open-sans": "^4.5.14",
    "@fortawesome/fontawesome-free": "6.4.2",
    "@ng-select/ng-select": "^11.1.1",
    "@ngrx/effects": "^16.0.1",
    "@ngrx/router-store": "^16.0.1",
    "@ngrx/store": "^16.0.1",
    "@nguniversal/express-engine": "^16.2.0",
    "@spartacus/assets": "~6.7.0",
    "@spartacus/core": "~6.7.0",
    "@spartacus/setup": "~6.7.0",
    "@spartacus/storefront": "~6.7.0",
    "@spartacus/styles": "~6.7.0",
    "angular-oauth2-oidc": "^15.0.1",
    "bootstrap": "^4.6.2",
    "express": "^4.15.2",
    "i18next": "^23.7.6",
    "i18next-http-backend": "^2.4.2",
    "i18next-resources-to-backend": "^1.2.0",
    "ngx-infinite-scroll": "^16.0.0",
    "rxjs": "~7.8.0",
    "ts-loader": "^9.4.4",
    "tslib": "^2.3.0",
    "zone.js": "~0.13.0"
  },
  "devDependencies": {
    "@angular-devkit/build-angular": "^16.2.8",
    "@angular/cli": "^0.5.0",
    "@angular/compiler-cli": "^16.2.0",
    "@nguniversal/builders": "^16.2.0",
    "@types/express": "^4.17.0",
    "@types/jasmine": "~4.3.0",
    "@types/node": "^16.11.7",
    "jasmine-core": "~4.6.0",
    "karma": "~6.4.0",
    "karma-chrome-launcher": "~3.2.0",
    "karma-coverage": "~2.2.0",
    "karma-jasmine": "~5.1.0",
    "karma-jasmine-html-reporter": "~2.1.0",
    "typescript": "~5.1.3"
  }
}"
`;

exports[`Spartacus Schematics: ng-add should add spartacus properly with SSR 3`] = `
"import { NgModule } from '@angular/core';
import { ServerModule } from '@angular/platform-server';

import { AppModule } from './app.module';
import { AppComponent } from './app.component';
import { provideServer } from '@spartacus/setup/ssr';

@NgModule({
  imports: [
    AppModule,
    ServerModule,
  ],
  bootstrap: [AppComponent],
  providers: [
    ...provideServer({
       serverRequestOrigin: process.env['SERVER_REQUEST_ORIGIN'],
     }),
  ],
})
export class AppServerModule {}
"
`;
